From 3329b195ceff7711560358f5beab15481064ddd9 Mon Sep 17 00:00:00 2001
From: Reilly Brogan <reilly@reillybrogan.com>
Date: Sat, 5 Aug 2023 12:44:18 -0500
Subject: [PATCH] Allow specifying the path to the gnome-session-ctl binary

Solus installs the gnome-session-ctl binary to a different path than the usual libexec directory.
---
 meson.build            | 2 ++
 meson_options.txt      | 1 +
 plugins/gsd.service.in | 2 +-
 plugins/meson.build    | 1 +
 4 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/meson.build b/meson.build
index 71f7ce3b..3dcf0e08 100644
--- a/meson.build
+++ b/meson.build
@@ -22,6 +22,7 @@ gsd_datadir = join_paths(gsd_prefix, get_option('datadir'))
 gsd_includedir = join_paths(gsd_prefix, get_option('includedir'))
 gsd_libdir = join_paths(gsd_prefix, get_option('libdir'))
 gsd_libexecdir = join_paths(gsd_prefix, get_option('libexecdir'))
+gsd_session_ctl_dir = join_paths(gsd_prefix, get_option('session_ctl_dir'))
 gsd_localedir = join_paths(gsd_prefix, get_option('localedir'))
 gsd_sysconfdir = join_paths(gsd_prefix, get_option('sysconfdir'))
 
@@ -265,6 +266,7 @@ output += '        prefix:                   ' + gsd_prefix + '\n'
 output += '        exec_prefix:              ' + gsd_prefix + '\n'
 output += '        libdir:                   ' + gsd_libdir + '\n'
 output += '        libexecdir:               ' + gsd_libexecdir + '\n'
+output += '        session_ctl_dir:          ' + gsd_session_ctl_dir + '\n'
 output += '        bindir:                   ' + gsd_bindir + '\n'
 output += '        sysconfdir:               ' + gsd_sysconfdir + '\n'
 output += '        datadir:                  ' + gsd_datadir + '\n\n'
diff --git a/meson_options.txt b/meson_options.txt
index a5eb354f..fa8549cd 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -1,5 +1,6 @@
 option('nssdb_dir', type: 'string', value: '', description: 'Absolute path to the system NSS database directory')
 option('udev_dir', type: 'string', value: '', description: 'Absolute path of the udev base directory')
+option('session_ctl_dir', type: 'string', value: '', description: 'Absolute path to the directory containing the gnome-session-ctl binary')
 option('systemd', type: 'boolean', value: true, description: 'Enable systemd integration')
 
 option('alsa', type: 'boolean', value: true, description: 'build with ALSA support (not optional on Linux platforms)')
diff --git a/plugins/gsd.service.in b/plugins/gsd.service.in
index c47a6768..8939ea4e 100644
--- a/plugins/gsd.service.in
+++ b/plugins/gsd.service.in
@@ -23,4 +23,4 @@ BusName=@plugin_dbus_name@
 TimeoutStopSec=5
 # We cannot use OnFailure as e.g. dependency failures are normal
 # https://github.com/systemd/systemd/issues/12352
-ExecStopPost=@libexecdir@/gnome-session-ctl --exec-stop-check
+ExecStopPost=@session_ctl_dir@/gnome-session-ctl --exec-stop-check
diff --git a/plugins/meson.build b/plugins/meson.build
index 16397dc6..06249cbf 100644
--- a/plugins/meson.build
+++ b/plugins/meson.build
@@ -119,6 +119,7 @@ foreach plugin: all_plugins
         unit_conf.set('plugin_name', plugin_name)
         unit_conf.set('description', plugin_description)
         unit_conf.set('libexecdir', gsd_libexecdir)
+        unit_conf.set('session_ctl_dir', gsd_session_ctl_dir)
         unit_conf.set('plugin_dbus_name', plugin_dbus_name)
 
         gates_all = []
-- 
2.41.0

